giraph-core/src/main/java/org/apache/giraph/writable/kryo/GiraphClassResolver.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.giraph.writable.kryo;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.KryoException;
+import com.esotericsoftware.kryo.Registration;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.util.DefaultClassResolver;
+import com.esotericsoftware.kryo.util.ObjectMap;
+import org.apache.giraph.zk.ZooKeeperExt;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooDefs;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+
+import static com.esotericsoftware.kryo.util.Util.getWrapperClass;
+
+/**
+ * This class resolver assigns unique classIds for every class that was not
+ * explicitly registered. It uses zookeeper for consistent mapping across all
+ * nodes.
+ */
+public class GiraphClassResolver extends DefaultClassResolver {
+  /** Length of the ZK sequence number */
+  private static final int SEQUENCE_NUMBER_LENGTH = 10;
+  /** Base ID to start for class name assignments.
+   * This number has to be high enough to not conflict with
+   * explicity registered class IDs.
+   * */
+  private static final int BASE_CLASS_ID = 10000;
+
+  /** Class name to ID cache */
+  private static Map<String, Integer> CLASS_NAME_TO_ID = new HashMap();
+  /** ID to class name cache */
+  private static Map<Integer, String> ID_TO_CLASS_NAME = new HashMap();
+  /** Zookeeper */
+  private static ZooKeeperExt ZK;
```

Comment:
Without static, we have to pass this information as a parameter. But there are so many places where the kryo objects are created, so it would not be pretty to delegate this information from BspService to all these places.